### ISSUE TITLE
API: Consider deprecation/exclusion of `GetQubitsAvailableTo{ Borrow | Use }()`

### ISSUE THREAD

**Issue by kuzminrobin** [microsoft/QuantumLibraries#418]
(Links [GetQubitsAvailableToBorrow()](https://docs.microsoft.com/en-us/qsharp/api/qsharp/microsoft.quantum.environment.getqubitsavailabletoborrow), [GetQubitsAvailableToUse()](https://docs.microsoft.com/en-us/qsharp/api/qsharp/microsoft.quantum.environment.getqubitsavailabletouse))

**Discussion** (key items)**:**

[12:25 PM] **Bettina Heim**
Hi all, 
We are in the process of adding the necessary QIR runtime functionality to support Q#. As part of that, I wanted to gather your insights regarding whether and how QubitsAvailableToBorrow and QubitsAvailableToUse are used. Right now, I am aware that the C# tracer uses them, but I am not sure anything else does. Are you aware of where else these may be used? Are these useful to keep or are these something that we might use internally only as part of the simulator but not necessarily expose in Q#? [12:27 PM] **Chris Granade** (THEY/THEM)
I don't know of anything else that uses them, no, but in part that's a bit circular. Most simulators don't suppose those operations, such that the utility for users is somewhat limited.
[12:28 PM] **Bettina Heim**
The primary use I see is for quick experimentation, but maybe there are better ways to do that without exposing them in Q#.
[12:29 PM] **Chris Granade** (THEY/THEM)
I'd be in support of deprecating and removing; an eventual application (e.g.: dynamic or virtual qubits — very cool stuff if you haven't seen!) would benefit from a better implementation anyway.
[12:34 PM] **Stefan Wernli**
I too would be in favor of simplifying by removing them. QIR simulation is unlikely to use the same qubit manager design as C#, and as of now there is no partner hardware or planned first party simulator that would use it.
[6:28 PM] **Chris Granade** (THEY/THEM)
As that would be a change to the API surface, we will want to make sure it's reviewed in the Q# API Review process so that the community has a chance to see our reasoning and to offer feedback. Robin, would you be willing to file a quantumlibraries issue, so that I can add it to our scheduling board? The next review is Thursday am, should be able to decide on it at that point. In the meantime, I think we can still proceed as-is under the understanding we may need to revert PRs should the change not be approved in review.

**Comment by cgranade** [comment]
To clarify, is the proposal to remove `GetQubitsAvailableToBorrow` and `GetQubitsAvailableToUse`?

**Comment by kuzminrobin** [comment]
Yes, based on the offline discussion the proposal is to remove GetQubitsAvailableToBorrow() and GetQubitsAvailableToUse().

**Comment by cgranade** [comment]
As discussed at https://github.com/microsoft/QuantumLibraries/pull/426, this proposal is approved. Subject to our Q# API design principles, to remove these operations we'll first need to deprecate them using the `@Deprecated("")` attribute. Thanks for driving this!

**Comment by cgranade** [comment]
> @bettinaheim, what are the next steps for me?

Thankfully, this one is a fairly straightforward proposal to implement, as it consists of adding `@Deprecated("")` to each deprecated operation, and adapting the API documentation comment to explain that the operations have been removed. I'm happy to do that, or for you to open that PR, either way; either let me know or assign yourself to mark that. I've also added this to the March milestone to indicate that we'll be doing this work in the March release.
